### ISSUE TITLE
Implement switching CurveSamplingQuality for Canvas/Workspace/Complex

### ISSUE THREAD

**Issue by dalboris** [vgc/vgc#1607]
Also see: #1602

Currently, changing the "sampling quality" (shortcut `i` when the canvas has focus) causes the following crash:

```
vgc.canvas: Switched edge subdivision quality to: Uniform Very High
libc++abi: terminating due to uncaught exception of type vgc::core::LogicError: Cannot perform the requested operation without an open undo group.
```

With the following stack trace:

![image](https://github.com/vgc/vgc/assets/4809739/032669a8-12b3-48bc-b5c3-4dd5e304118b)

As we can see, the problem is that:
- Changing the sampling quality is deferred to `Canvas::onPaintDraw()` [1]
- Which calls `ops::setKeyEdgeStrokeSamplingQuality(edge, quality)` for each edge
- Which emits the `Complex::nodesChanged()` signal with the edge having a `GeometryChanged` flag, because sampling quality is currently considered to be "authored geometry"
- Which causes an update from vacomplex to Workspace and DOM: `Workspace::updateElementFromVac_()` is called, eventually calling `setAttribute("positions"/"widths")` in the DOM, even though they haven't changed.
- Which causes the DOM to (rightfully) complain that it cannot be changed in this context (no UndoGroup)

[1] Why did we choose to defer? I think because we had to be in a graphics context to create/destroy some `graphics::GeometryView`. But maybe there's something to be improved there too: for example we could change the vacomplex sampling quality right away, and only defer updating the workspace graphics resources.

So the important question is: do we really want "sampling quality" to be part of "authored geometry". What to change if the answer is yes, and what to change if the answer is no?

As far as vacomplex is concerned, it does makes sense to consider this to be an actual change of the geometry, since it affects for example topological heuristics (e.g., which direction to use when gluing two edges?). And it makes the design simpler to only have `GeometryChanged` vs. both `GeometryChanged` and `MeshChanged`. In theory, the vacomplex doesn't know about UI, and doesn't care whether a DOM undo group should be opened or not when its geometry changes. Also note that currently, `MeshChanged` is not very useful, since it's only set when `GeometryChanged` is also set anyway.

However, it would also make a lot of sense if "sampling quality" wasn't considered a change of authored geometry. It can be seen as just a meta-setting that we need to set for computing heuristics that require some geometric computations, not an actual change of the vacomplex. And indeed, client code can decide to call `KeyEdgeData::computeStrokeSampling(quality)` with a different quality, and therefore display it differently. For performance reasons, it does makes sense to have the "sampling quality of the vacomplex" be the same as the "sampling quality of the Canvas", so that the sampling can be computed only once and shared between vacomplex and workspace. But semantically, they do not have to be the same, and it's up to the workspace to choose whether to take advantage of such sharing when possible. For example, if there are several canvases with different sampling quality, such sharing may not make sense. Or there could be a "preview quality" and "final render quality", etc.

Note that at the moment, sampling quality is never saved to the DOM. However, in the future, we might decide to actually make it a style attribute, set for the whole document, or individual cells. If such authoring of the DOM is allowed, then it would actually make sense that such change should be "undoable". But for now, since sampling quality is not undoable, I think it is reasonable to keep it simple and not consider it as a change of authored geometry, since this is out it is experienced by users.

We could later add it back as an actual authored changed, for example via a "CellProperty", which is now the preferred method to store style-related attributes.

**Comment by dalboris** [comment]
The PR #1610 fixes the crash by simply not emitting GeometryChange when the `vacomplex::KeyEdge` sampling quality changes.

Also, the sampling quality is now changed directly when the action in triggered, instead of being deferred to `onPaintDraw()`, since after analysis I concluded that it was safe to do so. Indeed, this calls `workspace::Edge::dirtyPreJoinGeometry_()` under the hood, which clears a bunch of graphics resources, but doing so is perfectly safe, since the cleared resources are simply moved to a "garbage list" and only actually released by the render thread at a later moment.

**Comment by dalboris** [comment]
However, I noticed another issue with the sampling quality (unrelated to the above fix): the chosen sampling quality (as stored in `Canvas::strokeSamplingQuality_` and modified by pressing the `i` key) is completely ignored when opening a new file, or copy-pasting objects (which is done via the DOM).

Indeed, `workspace::VacKeyEdge::updateFromDom_()` does call `vacomplex::ops::setKeyEdgeStrokeSamplingQuality(ke, strokeSamplingQuality_);` when it creates a new `vacomplex::KeyEdge`, but the data member `workspace::VacKeyEdge::strokeSamplingQuality_` is still initialized to its default value and never set to either the `Canvas` sampling quality or the `Workspace` default sampling quality.
